Document kc and k8s version constraints

pivotaljohn opened the pull request.

jtigger committed:
Document kc and k8s version constraints

jtigger committed:
Document component info fields in Downward API

joe-kimmel-vmw commented on site/content/kapp-controller/docs/develop/app-spec.md:
```diff
@@ -220,6 +220,16 @@ spec:
+              - name: kubeVersion
+                kubernetesVersion: {}
+              # query for all group versions (a list of the API groups and versions in the form "group/version") supported on this cluster
+              - name: apiVersions
```
here it would be slightly more clear to say something like `apiGVs`  or `apiGroupsNRoses`  or anyways to include a suggestion in the name that it's both groups and versions.

pivotaljohn replied:
my thought process was that it's like `kubectl api-versions`; the output of that command is literally the same thing. I was also trying to convey that this is a user-authored string... so I tried to make it different than the Downward API field/key.

joe-kimmel-vmw replied:
ah, oh ... it's not our bad name.   So maybe that'll be least-surprising to people who are already well-versed in k8s lore.   wait - is that really our target user?

pivotaljohn replied:
spoke about this out-of-band. I'm sold on the idea of keeping it very very straight-forward. 👍🏻

joe-kimmel-vmw commented on site/content/kapp-controller/docs/develop/packaging.md:
```diff
@@ -380,6 +380,73 @@ spec:
+
+### Requiring specific versions of kapp-controller
+
+A package may rely a feature of either kapp-controller itself or one of the bundled tools (i.e. ytt, kbld, helm, ...).
```
maybe (idk, it's so hard to explain) maybe like, "or one of the bundled tools used to fetch, template, or deploy"

joe-kimmel-vmw commented on site/content/kapp-controller/docs/develop/packaging.md:
```diff
@@ -380,6 +380,73 @@ spec:
+_(see [Package Consumer Concepts > Constraints](package-consumer-concepts.md#constraints) for syntax.)_
+
+If a PackageInstall resource is created for that package and the version of kapp-controller on that cluster falls outside 
+the given constraint, the installation will fail.
```
unless the version of kapp controller is so old[1] that it doesn't know how to interpret the constraint, in which case it will _ignore_ the constraint and the installation will succeed.  haha you're welcome.


[1] currently 100% of all deployed kapp-controllers are "so old"

joe-kimmel-vmw reviewed (commented):
hey thanks for figuring all this out.   I've picked a couple nits and there's probably an important corner case that will actually apply to the vast majority of users for the next ~6-12 months.

jtigger committed:
Include warning that older kc's won't constrain installs

- and be explicit about which version this feature is added for good
  measure.
- and make the variable that Kubernetes APIs populates obviously the
  "Groups and Version" listing.
